versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,416 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.bson.Document;
+import org.bson.types.Binary;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.ReferenceConflictException;
+import org.projectnessie.versioned.persist.adapter.CommitLogEntry;
+import org.projectnessie.versioned.persist.adapter.KeyList;
+import org.projectnessie.versioned.persist.adapter.KeyListEntity;
+import org.projectnessie.versioned.persist.adapter.KeyWithType;
+import org.projectnessie.versioned.persist.adapter.spi.DatabaseAdapterUtil;
+import org.projectnessie.versioned.persist.nontx.NonTransactionalDatabaseAdapter;
+import org.projectnessie.versioned.persist.nontx.NonTransactionalOperationContext;
+import org.projectnessie.versioned.persist.serialize.AdapterTypes.GlobalStateLogEntry;
+import org.projectnessie.versioned.persist.serialize.AdapterTypes.GlobalStatePointer;
+import org.projectnessie.versioned.persist.serialize.ProtoSerialization;
+
+public class MongoDatabaseAdapter
+    extends NonTransactionalDatabaseAdapter<MongoDatabaseAdapterConfig> {
+
+  private static final String GLOBAL_POINTER = "global_pointer";
+  private static final String GLOBAL_LOG = "global_log";
+  private static final String COMMIT_LOG = "commit_log";
+  private static final String KEY_LIST = "key_list";
+
+  private final ByteString keyPrefix;
+  private final byte[] globalPointerKey;
+
+  private final MongoClient mongoClient;
+  private final MongoCollection<Document> globalPointers;
+  private final MongoCollection<Document> globalLog;
+  private final MongoCollection<Document> commitLog;
+  private final MongoCollection<Document> keyLists;
+
+  protected MongoDatabaseAdapter(MongoDatabaseAdapterConfig config) {
+    super(config);
+
+    keyPrefix = ByteString.copyFromUtf8(config.getKeyPrefix());
+    globalPointerKey = keyPrefix.toByteArray();
+
+    ConnectionString cs = new ConnectionString(config.getConnectionString());
+    MongoClientSettings settings = MongoClientSettings.builder().applyConnectionString(cs).build();
+
+    mongoClient = MongoClients.create(settings);
+    MongoDatabase database = mongoClient.getDatabase(config.getDatabaseName());
+    globalPointers = database.getCollection(GLOBAL_POINTER);
+    globalLog = database.getCollection(GLOBAL_LOG);
+    commitLog = database.getCollection(COMMIT_LOG);
+    keyLists = database.getCollection(KEY_LIST);
+  }
+
+  @Override
+  public void reinitializeRepo(String defaultBranchName) {
+    globalPointers.deleteMany(Filters.empty()); // empty filter matches all
+    globalLog.deleteMany(Filters.empty());
+    commitLog.deleteMany(Filters.empty());
+    keyLists.deleteMany(Filters.empty());
+
+    super.initializeRepo(defaultBranchName);
+  }
+
+  @Override
+  public void close() {
+    mongoClient.close();
+  }
+
+  private byte[] toId(Hash id) {
```

Comment:
Represent hashes as Strings in Mongo?

Reply:
Sure, but I think it's going to be more processing work with the current `persist.proto` objects :thinking: I'll give it a try.

Reply:
It's really just about representing it using the types available in mongo. (note: the jdbc-like adapters use `VARCHAR` as well).

Reply:
Mongo uses binary serialization internally, AFAIK, and supports byte arrays, but I see the rationale for using a human-readable type for keys :+1:

Reply:
Reworked to `String` in 1f3f1e4c3afa0f1266f7906e09a58fb549318f66